wp-includes/class-wp-service-workers.php

```diff
@@ -0,0 +1,224 @@
+	 * @param array           $deps   Optional. An array of registered item handles this item depends on. Default empty array.
+	 * @param array           $scopes Optional. Scopes of the service worker. Default relative path.
+	 * @return bool Whether the item has been registered. True on success, false on failure.
+	 */
+	public function register( $handle, $src, $deps = array(), $scopes = array() ) {
+
+		// Set default scope if missing.
+		if ( empty( $scopes ) ) {
+			$scopes = array( site_url( '/', 'relative' ) );
+		}
+		return parent::add( $handle, $src, $deps, false, compact( 'scopes' ) );
+	}
+
+	/**
+	 * Get service worker logic for scope.
+	 *
+	 * @param string $scope Scope of the Service Worker.
+	 */
+	public function serve_request( $scope ) {
+
+		header( 'Content-Type: text/javascript; charset=utf-8' );
+		nocache_headers();
+
+		if ( ! in_array( $scope, $this->get_scopes(), true ) ) {
+			status_header( 404 );
+			return;
+		}
+
+		$scope_items = array();
+
+		// Get handles from the relevant scope only.
+		foreach ( $this->registered as $handle => $item ) {
+			if ( in_array( $scope, $item->args['scopes'], true ) ) {
+				$scope_items[] = $handle;
+			}
+		}
+
+		$this->output = '';
+		$this->do_items( $scope_items );
+
+		$file_hash = md5( $this->output );
+		header( "Etag: $file_hash" );
+
+		$etag_header = isset( $_SERVER['HTTP_IF_NONE_MATCH'] ) ? trim( $_SERVER['HTTP_IF_NONE_MATCH'] ) : false;
+		if ( $file_hash === $etag_header ) {
+			status_header( 304 );
+			return;
+		}
+
+		// @codingStandardsIgnoreLine
+		echo $this->output;
+	}
+
+	/**
+	 * Get all scopes.
+	 *
+	 * @return array Array of scopes.
+	 */
+	public function get_scopes() {
+
+		$scopes = array();
+		foreach ( $this->registered as $handle => $item ) {
+			$scopes = array_merge( $scopes, $item->args['scopes'] );
+		}
+		return array_unique( $scopes );
+	}
+
+	/**
+	 * Process one registered script.
+	 *
+	 * @param string $handle Handle.
+	 * @param bool   $group Group.
+	 * @return void
+	 */
+	public function do_item( $handle, $group = false ) {
+		global $wp_filesystem;
+
+		$obj = $this->registered[ $handle ];
+
+		if ( is_callable( $obj->src ) ) {
+			$this->output .= call_user_func( $obj->src ) . "\n";
+		} else {
+			$validated_path = $this->get_validated_file_path( $obj->src );
+			if ( is_wp_error( $validated_path ) ) {
+				_doing_it_wrong(
+					__FUNCTION__,
+					/* translators: %s is file URL */
+					sprintf( esc_html__( 'Service worker src is incorrect: %s', 'pwa' ), esc_html( $obj->src ) ),
+					'0.1'
+				);
+
+				/* translators: %s is file URL */
+				$this->output .= "console.warn( '" . sprintf( esc_html__( 'Service worker src is incorrect: %s', 'pwa' ), esc_html( $obj->src ) ) . "' );\n";
+			} else {
+				/* translators: %s is file URL */
+				$this->output .= sprintf( esc_html( "\n/* Source: %s */\n" ), esc_url( $obj->src ) );
+				$this->output .= "( function() {\n";
```

Comment:
On second thought, this IIFE may cause confusion since it is a different behavior than WordPress normally handles scripts. If a service worker script doesn't want to conflict with others, than it shod wrap itself in an IIFE, no?

Reply:
Makes sense!